library/psa_crypto_pake.c

```diff
@@ -0,0 +1,720 @@
+        }
+    }
+    else if( operation->state == PSA_PAKE_INPUT_X4S &&
+             operation->sequence == PSA_PAKE_X1_STEP_ZK_PROOF )
+    {
+        ret = mbedtls_ecjpake_read_round_two( &operation->ctx.ecjpake,
+                                              operation->buffer,
+                                              operation->buffer_length );
+
+        mbedtls_platform_zeroize( operation->buffer, PSA_PAKE_BUFFER_SIZE );
+        operation->buffer_length = 0;
+
+        if( ret != 0 )
+        {
+            psa_pake_abort( operation );
+            return( mbedtls_to_psa_error( ret ) );
+        }
+    }
+
+    if( ( operation->state == PSA_PAKE_INPUT_X1_X2 &&
+          operation->sequence == PSA_PAKE_X2_STEP_ZK_PROOF ) ||
+        ( operation->state == PSA_PAKE_INPUT_X4S &&
+          operation->sequence == PSA_PAKE_X1_STEP_ZK_PROOF ) )
+    {
+        operation->state = PSA_PAKE_STATE_READY;
+        operation->input_step++;
+        operation->sequence = 0;
+    }
+    else
+    {
+        operation->sequence++;
+    }
+
+    return( PSA_SUCCESS );
+#else
+    return( PSA_ERROR_NOT_SUPPORTED );
+#endif
+}
+
+psa_status_t psa_pake_get_implicit_key(psa_pake_operation_t *operation,
+                                       psa_key_derivation_operation_t *output)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    if( operation->alg == 0 ||
+        operation->state != PSA_PAKE_STATE_READY ||
+        ( operation->input_step != PSA_PAKE_STEP_DERIVE  &&
+          operation->output_step != PSA_PAKE_STEP_DERIVE ) )
+        return( PSA_ERROR_BAD_STATE );
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_ECJPAKE)
+    ret = mbedtls_ecjpake_derive_secret( &operation->ctx.ecjpake,
+                                         operation->buffer,
+                                         PSA_PAKE_BUFFER_SIZE,
+                                         &operation->buffer_length,
+                                         mbedtls_psa_get_random,
+                                         MBEDTLS_PSA_RANDOM_STATE );
+    if( ret != 0)
+    {
+        psa_pake_abort( operation );
+        return( mbedtls_to_psa_error( ret ) );
+    }
+
+    status = psa_key_derivation_input_bytes( output,
+                                             PSA_KEY_DERIVATION_INPUT_SECRET,
+                                             operation->buffer,
+                                             operation->buffer_length );
+
+    mbedtls_platform_zeroize( operation->buffer, PSA_PAKE_BUFFER_SIZE );
+
+    psa_pake_abort( operation );
+
+    return( status );
+#else
+    return( PSA_ERROR_NOT_SUPPORTED );
+#endif
+}
+
+psa_status_t psa_pake_abort(psa_pake_operation_t * operation)
+{
+    if( operation->alg == 0 )
+    {
+        return( PSA_SUCCESS );
+    }
+
+    operation->alg = 0;
+    operation->state = 0;
+    operation->sequence = 0;
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_ECJPAKE)
```

Comment:
Again, shouldn't we check `operation->alg`? Most of these assignments are safe, but the call to `mbedtls_ecjpake_free()` seems dangerous if the `ctx` union holds something other than an ecjpake context. (Again, the code is correct now, but there's a risk of bugs here when we introduce more PAKE algs in the future.)